### ISSUE TITLE
idea: new inheritance model

### ISSUE THREAD

**Issue by vtjnash** [JuliaGraphics/Gtk.jl#384]
we want to represent inheritance as something rather abstract, such as `GLeaf{:GButton} <: GInterface{:GButton} <: GInterface{:GBox} <: GObject`, but that's not directly possible, so instead we currently create a mirror type hierarchy (to make it possible to use Glib types in dispatch in Julia). However, our current approach has some rough edges—notably in the macro that generates them. One thought I had recently is to restructure the types to use some amount of parameterization to switch around the ownership model.

```julia
abstract GObject # common supertype of all
struct GBox <: GObject end # these could have extra fields
struct GButton <: GBox end

abstract GAbstract{T} # refer to this type to permit inheritance
struct GLeaf{T} <: GAbstract{T} # refer to this type to restrict to the actual instance
    handle::Ptr{Void}
    flds::T
end

# typealiases to the current interface
const GObject = GAbstract{<:GObject}
const GBox = GAbstract{<:GBox}
const GButton = GAbstract{<:GButton}

# typealiases to the current interface
const GButtonLeaf = GLeaf{GButton}
```
```
# some usage, mostly unchanged
GButton(args...; kwargs...) = GButtonLeaf(args...; kwargs...) # forward interface constructor to a concrete instance of it
f(::GBox) = 1 # dispatch on subtypes of GBox
f(::GButton) = 2
```

**Comment by tknopp** [comment]
you mean
```
struct GBox <: GObject end # these could have extra fields
struct GButton <: GObject end
```
right? In you code GButton inherits from GBox, which is not possible.

**Comment by tknopp** [comment]
I cannot comment to much on your plan but would like to showcase one thing feels wrong in the current system. Suppose I have a `list::ListStore` and I want to call `G_.sort_column_id` then I need to explicitly cast to `TreeSortable` like this
```
G_.sort_column_id(TreeSortable(list),0,GtkSortType.ASCENDING)
```
This is really not cool and also not necessary in C/Python. I know that in a previous iteration of the Gtk type system this was not necessary. Will your proposal solve this?

**Comment by vtjnash** [comment]
> In you code GButton inherits from GBox, which is not possible

Oh, oops. You're right. Ugggh, maybe this really isn't possible to do well.

> this was not necessary

I don't recall that. It actually is necessary in C (this function would be `gtk_tree_sortable_get_sort_column_id`) but it's much less obvious, since everything works that way. My proposal above wouldn't help with that.

The problem is that we can only represent the concrete inheritance tree in the Julia type system, (whereas python has multiple-inheritence, so it can also represent the inherited interfaces). In Julia, we usually represent interfaces with traits (currently, "can be converted to")—but the impedance mismatch with glib types is rather high in that case and I don't have a good solution to this. We could stop declaring custom Julia types, and then implement our own OO-type system in the methods that care about it, but I'm not sure that's such a great solution either:
```julia
struct GObject{T}
    handle::Ptr{Void}
    fields::T
end
struct NoFields{GName::Symbol}; end
struct GtkBoxFields; end

# typealiases to the current interface
const GObject = GObject{NoFields{:Object}}
const GtkBox = GObject{NoFields{:GtkBox}}
const GtkButton = GObject{NoFields{:GtkButton}}
const CustomGBox = GObject{GtkBoxFields}

# some usage is very different
f(g::GObject) =
    g_isa(g, :GBox) ? 1 : (g_isa(g, :GButton) ? 2 : error()) # "dispatch" on subtypes of GBox
f(::CustomGBox) = 3

**Comment by tknopp** [comment]
> I don't recall that. 

That is fortunately well documented ;-)

https://github.com/JuliaGraphics/Gtk.jl/issues/89

Maybe we were ducktyping before. I also recall that I once created Unions for interfaces.

**Comment by vtjnash** [comment]
Ah, yes, documentation is good for keeping me honest. So looking at https://github.com/JuliaGraphics/Gtk.jl/commit/fc7c282867982fb4768460fc4a9496c34f63f788, it looks like I/We had previously mapped each Interface to exactly one of its concrete implementations (or in this particular case, exactly two). This meant you could use `sort_column_id(list, 0, GtkSortType.ASCENDING)` to sort any `list::TreeSortable`, but only if your implementation of a `TreeSortable` was one of precisely `Union(GtkListStore, GtkTreeStore)`

**Comment by tknopp** [comment]
It was not one but it could be many. The issue was of course extensibility although practically this is not so much a concern if we cover all types included within Gtk.

What we truly need is abstract multiple inheritance which (hopefully) in some form will be developed for julia 2.0. I have no experience with traits but the promises always have been that they allow emulating multiple inheritance. 

@mauro3: Is that correct that one can emulate multiple inheritance with Traits.jl? We want to express 
```
GtkListStore <: GObject
GtkListStore <: GtkTreeStore
```

**Comment by mauro3** [comment]
Yes, probably you can in principle, but I'm not aware of any easy way to do it at the moment. The approach Traits.jl took does work anymore.

**Comment by timholy** [comment]
It's not hard to do manually:
```julia
abstract type GObjectKind end
struct IsGObject <: GObjectKind end
struct IsNotGObject <: GObjectKind end
GObjectKind(obj) = IsNotGObject()    # by default, things are not GObjects

abstract type GtkTreeStoreKind end
struct IsGtkTreeStore <: GtkTreeStoreKind end
struct IsNotGtkTreeStore <: GtkTreeStoreKind end
GtkTreeStoreKind(obj) = IsNotGtkTreeStore()   # by default, things are not GtkTreeStores

# But a GtkListStore is "true" for both
GObjectKind(::GtkListStore) = IsGObject()
GtkTreeStoreKind(::GtkListStore) = IsGtkTreeStore()

# The "depot function" determines the traits of the inputs
foo(obj, args...) = _foo(GObjectKind(obj), GtkTreeStoreKind(obj), obj, args...)
# Several specific implementations dispatch on the trait values rather than the object itself
_foo(::IsGObject, ::IsGtkTreeStore, obj, args...) = # implementation
```

But a traits package may come in the next couple of months, see https://github.com/andyferris/Traitor.jl/issues/8. https://github.com/JuliaLang/julia/pull/28916 should help a lot.

**Comment by tknopp** [comment]
The trick is that foo is duck typed and then there is some manual dispatching going on. That seems like a very good approach for Interfaces like `TreeSortable`. The other question is if Traits could also help with the mirrored type hierarchy that we currently have. Currently we have an abstract type hierarchy that mirrors Gtk ones but if you want a concrete type, we have the `Leaf` version that branches of the abstract type hierarchy. This was the original intention from Jameson here.
